### DISCUSSION TITLE
How to simulate interactions between pairwise GSMs?

### DISCUSSION THREAD

**Discussion by EmmettPeng** [micom-dev/micom#71] [Q&A] [answered]
Hi micom creators,

Thanks for developing this great tool. I have assembled several MAGs from my metagenome sequencing data and built GSMs using CarveMe tool for each MAG. Here, I intended to predict the potential interactions between each two GSMs. I have tried micom community pFBA.

First, two GSMs were loaded and the growth rate were calculated using `optimize()`. Then I used `micom.Community()` function to create a community made up of these two taxa of equal abundance. Then I used cooperative trade-off results (for each GSM, the result was multiplied by 2 as its relative abundance is 0.5) to denote the optimal growth rate when two taxa grew together. 

What I found quite confusing is that for a given GSM (e.g. GSM A), no matter what the other GSM (GSM B, C, D...) was, the predicted growth rates of GSM A almost stayed the same. This was very strange because one should show different interactions with others. Since I am totally a stranger to FBA computations, I am hoping someone could explain how I should solve this problems and maybe offer some further advice.

Thanks again!

Emmett

**Comment by cdiener** [comment, marked as answer]
Hi, welcome to using MICOM :tada: 

We did notice that the results for co-cultures tend to depend a lot on relative abundances and the tradeoff parameters. In fact simulating co-cultures turns out to be more complicated than a full microbiota in some cases. So here is what we found works best for us:

1. Use a very high tradeoff (like 0.99 or even 1). Competition for 2 taxa always tends to be much less than for 100 taxa so higher tradeoff values tend to get better results.
2. For taxon-taxon interactions you may also want to try to run the "minimal imports" strategy. pFBA works for for internal fluxes but tends to work less well for transports or taxon-taxon interactions since those are often not really connected to an enzyme and the parsimonity assumption fails wo work. If you use the [grow workflow](https://micom-dev.github.io/micom/high_level.html#Simulating-growth) you can just pass `strategy="minimal imports"`. Or for a single model you can use `minimal_medium(..., solution=True)` (see [here](https://micom-dev.github.io/micom/autoapi/micom/media/index.html#micom.media.minimal_medium)). This often does a better job of identifying exchanges.
3. Make sure your assumptions are correct. If you set the relative abundances to 0.5 for each taxon that denotes the *steady state* abundance and not the inoculation abundance. So the abundance you would get after running the co-culture in a bioreactor for a long time. Those may differ from the inoculation abundance by quite a bit. So if you have evidence that one strain outcompetes the other in wet lab experiments that may clash with the 50:50 assumption and give you wrong predictions. If you used MAGs then using the relative abundances from the metagenomic samples may work better. But using all MAGs from a single sample with the corresponding abundances probably works best. MICOM tends to give better predictions the more taxa you include.
4. Make sure you apply the correct medium to the model before simulating and always gapfill carveme models on realistic growth conditions. CarveME does not actually use flux bounds in the media formulation so we recommend supplying a quantitative growth medium. What medium did you use here? 

You said that GSM A should have interactions with other taxa. What evidence is that based on? 

Finally, MICOM does scale all growth rates and fluxes correctly so you don't need to multiply anything by 2 or something. That will actually give you wrong results.

**Reply by EmmettPeng** [reply]
Thanks for your detailed reply! 

For your point 2, I agree that my workflow should take more of transport or exchange reactions into consideration because I intend to focus on the complementarity potential between these MAGs (for example, cross-feeding).

Actually, my data came from environmental samples, which makes it difficult to give a detailed medium composition. But, I only need to predict the potential interactions two MAGs might show, under ideal conditions. For instance, if A gives out a specific metabolite extracellular and B can uptake it as nutrition source, they should possess a complementarity potential, right? So under this circumstance, what medium should I apply?

**Reply by cdiener** [reply]
Okay that explains why you don't see interactions. If you don't provide a medium definition to carveme it will build a model where every possible metabolite that the organism is able to metabolize is present in the medium. So that organism can never depend on other organisms in that setting. When MICOM builds the community model it will use the superset of both media which will drown out any potential interactions since every possible metabolite any of the organism needs is already available in the medium. So you would need a more realistic growth medium to study those interactions. There are helper functions in micom (`complete_db_medium` for instance) that can help you if you have only limited information about the environment. Alternatively, you could also simulate them in a defined medium such as M9 etc.

**Comment by bananabenana** [comment]
Hi @cdiener , firstly I wanted to say what a brilliant tool you've developed here. I am enjoying digging into the many useful functions and applications.

I have a similar query. I am looking at two-strain co-cultures to test a few things, using whole genome models.
I am using a defined media, where isolate A can grow on the carbon source, while isolate B cannot. This has been verified using these models via FBA and lab experiments. 

```
tradeoff(manifest, model_folder="community_models", medium=medium, threads=4)
```

- When running the `tradeoff()` function and looking at the NaN tradeoff, I get close to what I expect, isolate A will produce biomass of ~2, while isolate B will be ~0.

- **From tradeoff 0.1 - 1, biomass for both isolates A and B are identical across each tradeoff value. They are the same species but different strains, with identical biomass functions. Does this seem correct?**

- In the case of some carbon sources which isolate B cannot use, but isolate A can, isolate B will produce ~0.6, while isolate A will produce ~1.4 biomass, which MICOM indicates is due to crossfeeding, as Strain A is exporting breakdown metabolites from the carbon source, which isolate B is uptaking.

- **Is this expected behaviour of MICOM tradeoff simulations, that biomass would be equal between isolates A and B?**
- I am wondering if this is a realistic scenario, that isolate A would uniquely have access to a carbon source, then export the breakdown components for community uptake. This can be tested in the lab, just wanting to find out before going to the effort!

```
grow(manifest, model_folder="community_models", medium = medium_read, tradeoff=0, threads=4)
```

- When I use the `grow()` function with `tradeoff=0`, which I thought would be consistent with `NaN` from `tradeoff()` I am getting very small growth of both isolates A and B, rather than the expected biomass of isolate A ~2 and isolate B ~0.

- **Is there any way to replicate the `NaN` result under `grow()`? To be clear, I do not want to necessarily optimise for growth of both strains, as this isn't expected behaviour. Am I using MICOM inappropriately here?**


```
# Build community
com = Community(community_data_reduced)

# Create custom media environment in which isolate B won't grow in isolation but A will
galt_dict = {'EX_ca2_m': 10, 'EX_cl_m': 10, 'EX_co2_m': 10, 'EX_cobalt2_m': 10, 'EX_cu2_m': 10, 'EX_fe2_m': 10, 'EX_fe3_m': 10, 'EX_galt_m': 20, 'EX_h_m': 10, 'EX_h2o_m': 10, 'EX_k_m': 10, 'EX_mg2_m': 10, 'EX_mn2_m': 10, 'EX_mobd_m': 10, 'EX_na1_m': 10, 'EX_nh4_m': 10, 'EX_ni2_m': 10, 'EX_o2_m': 20, 'EX_pi_m': 10, 'EX_so4_m': 10, 'EX_zn2_m': 10}

# Change medium of community
com.medium = galt_dict

# Optimise
com.optimize()

# Run cooperative tradeoff
sol_galt = com.cooperative_tradeoff(fluxes=True, pfba=True, min_growth=0)

# Look at objective value of members
sol_galt.members
```

- When using the Community() approach with a custom media, growth rates were once again matched between A and B. 
- Is this expected behaviour? If so, is there a function in MICOM which will allow me to look at head to head growth comparison of isolates A and B in the absence of crossfeeding?

**Reply by cdiener** [reply]
Hi,

I think the crux is in point 3 of the previous reply, specifically:

> Make sure your assumptions are correct. If you set the relative abundances to 0.5 for each taxon that denotes the steady state abundance and not the inoculation abundance. So the abundance you would get after running the co-culture in a bioreactor for a long time. Those may differ from the inoculation abundance by quite a bit. So if you have evidence that one strain outcompetes the other in wet lab experiments that may clash with the 50:50 assumption and give you wrong predictions. If you used MAGs then using the relative abundances from the metagenomic samples may work better. But using all MAGs from a single sample with the corresponding abundances probably works best. MICOM tends to give better predictions the more taxa you include.

Basically, you are telling MICOM that you think that strain A and B have the same non-zero abundance in the exponential growth phase (steady state assumption with growth is only valid there). Since isolate B is at the same abundance as A, it is very unlikely it can't grow while B can, so MICOM will predict growth for both. This does not have to be exactly the same growth rate, but it will be in very rich media like yours (not necessarily rich in no. components but large fluxes). So if you want MICOM to predict somewhat more realistic fluxes, you will need to get the abundances in log-phase. You can simulate using the inoculation as abundances and get useful insights, but this depends on how much that deviates from the real log-phase abundances.

In terms of the cross-feeding I would say that this is at least pretty common. One question is whether the secretion of breakdown products is optional for strain A or obligate. Running a pFBA followed by an FVA on those secreted compounds should give you some evidence whether those will always occur when A growth. *If* they do, that probably means that B would be able to grow in the presence of A since in the initial inoculum A can't outcompete B yet and will also provide the missing metabolites for B to grow, making B competitive and pushing it away from extinction.

Simulating in the absence of cross-feeding would be done by simply optimizing the strain models individually, however, that is probably not ecologically relevant. Strains grown together will interact and compete for resources and fitness. And exploiter-victim competition as you see it here is pretty common...

**Reply by bananabenana** [reply]
Thanks for the fast response.

> This does not have to be exactly the same growth rate, but it will be in very rich media like yours (not necessarily rich in no. components but large fluxes). 

Ahhh yep makes sense.

> One question is whether the secretion of breakdown products is optional for strain A or obligate. Running a pFBA followed by an FVA on those secreted compounds should give you some evidence whether those will always occur when A growth. 

Fantastic idea, I am slapping my head for not testing that.


> Simulating in the absence of cross-feeding would be done by simply optimizing the strain models individually, however, that is probably not ecologically relevant. 

Yep that makes sense. Agreed.


Okay thanks again for the feedback and for the great tool. You are making genuinely amazing advancements in this space.
